### PR TITLE
Code-behind fragment and widget access fixes

### PR DESCRIPTION
**Pull request by grendello**
Fragment handling code partially taken from https://github.com/xamarin/xamarin-android/pull/1302/

Fragments are treated as normal widget, with the difference that they have to be
found using `FragmentManager.FindFragmentById` instead of the usual
`Activity.FindViewById` method.

Each generated class also gets a property named `Widget` which returns the
actual Android widget as found in the layout file. This allows us to keep
hierarchical nature of the XML code (thus handling nested widgets with duplicate
ids gracefully) while being able to access the parent widget itself.

The commit introduces a new attribute called `tools:managedType` which is used
to specify the element associated property's type. We cannot use `tools:class`
for this purpose since the layout root element already uses it to specify the
name of the generated code-behind class and if the element had `android:id` on
it, we would end up using the activity's type for the root element's property in
the generated code, instead of its actual type.

### CODE REVIEW

**Comment by jonpryor** [comment]
Missing: Documentation update. ;-)

There's also a philosophical question which we've debated on (private?) Slack and Issue #1302: What API are we actually providing, and should this be *consistent*?

For example, if a `.axml` file has this fragment:

```xml
<TextView
    android:id="@+id/first_text_view"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:editable="false"
/>
```

What API are we exporting? There are three options, one of which *must be documented*:

**Option 1**: "Direct" binding- We bind as it is in the `.axml` file:

```csharp
internal TextView first_text_view {get;}
```

This affords the "obvious" usage:

```csharp
protected override void OnCreate (Bundle bundle) {
    first_text_view.Click += (o, e) => { /* ... */ };
}
```

However, this option won't -- can't! -- support `.axml` files which have multiple `//@android:id` elements with the same value. Presumably, we'd bind the first one found.

**Option 2**: "Indirect" binding: we bind as an indirect type in the `.axml` file:

```csharp
internal __first_text_view_Views first_text_view {get;}
internal class __first_text_view_Views {
    public TextView Value {get;} // name?
}
```

This doesn't support the obvious usage of **Option 1**, but means that if the `.axml` declares multiple elements with the same `//@android:id` attribute, we can list them all in a consistent manner.

Usage is...weird:

```csharp
protected override void OnCreate (Bundle bundle) {
    first_text_view.Value.Click += (o, e) => { /* ... */ };
    // WTF is `Value`? What do we call the "other" `//@android:id` "overloads"? Value2, Value3, etc.?
}
```

**Option 3**: Do both! Use **Option 1** when there's only one `//@android:id` value within the file, and **Option 2** otherwise.

This seems "obvious", but means there is no consistent API; it's dependent on how many `//@android:id` elements share the same value.

This *could* be acceptable -- maybe even preferable?! -- but will complicate the documentation process.

File: src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutCodeBehind.cs [hunk 1]
```diff
@@ -121,7 +122,7 @@ public override bool Execute ()
 
 		void GetLineInfo (IXmlLineInfo linfo, out int line, out int column)
 		{
-			if (linfo == null || linfo.HasLineInfo ()) {
+			if (linfo == null || !linfo.HasLineInfo ()) {
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutCodeBehind.cs]: I half wonder if this should be a separate PR.

**Comment by grendello** [comment]
For example, if a .axml file has this fragment:

<TextView
    android:id="@+id/first_text_view"
    android:layout_width="wrap_content"
    android:layout_height="wrap_content"
    android:editable="false"
/>

What API are we exporting? There are three options, one of which must be documented:

> Option 1: "Direct" binding- We bind as it is in the .axml file:
> 
> internal TextView first_text_view {get;}
> 
> This affords the "obvious" usage:
> 
> ```
> protected override void OnCreate (Bundle bundle) {
>   first_text_view.Click += (o, e) => { /* ... */ };
> }
>```
> However, this option won't -- can't! -- support .axml files which have multiple //@android:id elements with the same value. Presumably, we'd bind the first one found.

This option also doesn't take into account the hierarchy. I realize we can flatten it, but then you lose the contextual information from the axml and I think having this hierarchical context is, in fact, an advantage of our implementation over Android.

> Usage is...weird:
>
> ```
> protected override void OnCreate (Bundle bundle) {
>   first_text_view.Value.Click += (o, e) => { /* ... */ };
>    // WTF is `Value`? What do we call the "other" `//@android:id` "overloads"? Value2, Value3, etc.?
I don't know about "weird" here. This is used in all kinds of OOP GUI libraries where there's the notion of a container and its children or properties. As for the "other" overloads - the problem doesn't exist here because the other ones are nested in other parents and there's always unique and uniform path to access them:

```csharp
protected override void OnCreate (Bundle bundle) {
   // MyTextView is the .axml widget id
   first_layout.MyTextView.Widget.Click += (o, e) => { };
   second_layout.MyTextView.Widget.Click += (o, e) => { };
}
```
As for sibling widgets with the same id, they aren't directly accessible in Android either - the first one is picked and we bind that one. The other dupes are reachable by enumerating children of the layout/parent/container widget.
**Note**: I used `Widget` above as the commit implements because I think it better implements the "meaning" of the property.

>}
>
> Option 3: Do both! Use Option 1 when there's only one //@android:id value within the file, and Option 2 otherwise.
>
> This seems "obvious", but means there is no consistent API; it's dependent on how many //@android:id elements share the same value.
> 
> This could be acceptable -- maybe even preferable?! -- but will complicate the documentation process.

You seem to focus only on the duplicate IDs, but they are just part of the entire picture. I really think that the biggest advantage of my approach is the reflection of the hierarchical nature of the layout and it blends in nicely within the overall OOP paradigm of XA.

File: src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutCodeBehind.cs (hunk 1 above)
Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutCodeBehind.cs]: Nah. It's a one-liner and really doesn't deserve special treatment.

File: Documentation/guides/LayoutCodeBehind.md [hunk 2]
```diff
@@ -11,28 +11,45 @@ dateupdated: 2018-01-29
...
+```
 
 <a name="" class="injected"/></a>
 
 # Preparing to use Code Behind
 
 In order to make use of this new feature there are a few changes which are required. 
-An axml/xml file that you want to associate with an activity needs to be modified to 
-include a few extra xml attributes on the root layout element.
+An ``axml/xml`` file that you want to associate with an activity needs to be modified to
```
Comment (jonpryor) [inline Documentation/guides/LayoutCodeBehind.md]: Why the double backticks here?

File: Documentation/guides/LayoutCodeBehind.md [hunk 3]
```diff
@@ -11,28 +11,45 @@ dateupdated: 2018-01-29
...
 
 # Preparing to use Code Behind
 
 In order to make use of this new feature there are a few changes which are required. 
-An axml/xml file that you want to associate with an activity needs to be modified to 
-include a few extra xml attributes on the root layout element.
+An ``axml/xml`` file that you want to associate with an activity needs to be modified to 
+include a few extra xml attributes on the root layout element. 
+
+Additionally, **only*** elements which have the `android:id` attribute will be accessible via
```
Comment (jonpryor) [inline Documentation/guides/LayoutCodeBehind.md]: Too many -- or not enough -- `*`s. Presumably you want `***only***`.

File: Documentation/guides/LayoutCodeBehind.md [hunk 4]
```diff
@@ -79,6 +107,135 @@ situation in a manner which fits the app they are writing. For example
...
+The code-behind will have this rough structure (class names are different to keep the documentation clear):
+
+```csharp
+myScrollView_Class myScrollView { 
+   get { return new myScrollView_Class (this); }
+}
+
+class myScrollView_Class 
+{
+   public ScrollView Widget {
```
Comment (jonpryor) [inline Documentation/guides/LayoutCodeBehind.md]: In terms of keeping code short for clarity reasons, we should go with:

```csharp
public ScrollView Widget {get;}
```

File: Documentation/guides/LayoutCodeBehind.md [hunk 5]
```diff
@@ -79,6 +107,135 @@ situation in a manner which fits the app they are writing. For example
...
+```
+
+### Code structure rationale
+
+It may seem that it would be simpler to generate code which would put properties returning the layout elements directly in the
+Activity partial class instead of outputing a seemingly complex nested class structure. This approach would work if it wasn't
+for the following:
+
+  1. Android allows duplicate `android:id` values for **sibling** elements
+  2. Android allows duplicate `android:id` values anywhere withn the layout tree
```
Comment (jonpryor) [inline Documentation/guides/LayoutCodeBehind.md]: Spelling error: you want `within`.

File: Documentation/guides/LayoutCodeBehind.md [hunk 6]
```diff
@@ -79,6 +107,135 @@ situation in a manner which fits the app they are writing. For example
...
+  3. Many layouts reuse XML in the form of fragments
+  4. Many layouts reuse XML in the form of includes (using the `<include>` element)
+
+`1.` means that there's direct access (via `FindViewById` or with code-behind) to the **first** element with that `id` **only**.
```
Comment (jonpryor) [inline Documentation/guides/LayoutCodeBehind.md]: Instead of enclosing numbers in backticks, please surround with parens, e.g. **(1)**, **(2)**, etc.

File: src/Mono.Android/Test/Xamarin.Android.RuntimeTests/MainActivity.cs [hunk 7]
```diff
@@ -1,16 +1,25 @@
...
 			Name="xamarin.android.runtimetests.MainActivity")]
-	public class MainActivity : TestSuiteActivity
+	public partial class MainActivity : TestSuiteActivity
 	{
 		protected override void OnCreate (Bundle bundle)
 		{
+			// The actual view is set/replaced in `TestSuiteActivity.OnCreate()`
+			SetContentView (Resource.Layout.Main);
+
+			first_text_view.Widget.Click += delegate {
```
Comment (jonpryor) [inline src/Mono.Android/Test/Xamarin.Android.RuntimeTests/MainActivity.cs]: Question: `Main.axml` contains no `<include/>`s -- which aren't supported yet, anyway -- and only one element with `android:id="@ +id/first_text_view"`.

Why isn't this a *leaf* node? Why do we need the extra `.Widget` here?

File: Documentation/guides/LayoutCodeBehind.md (hunk 2 above)
Comment (grendello) [reply Documentation/guides/LayoutCodeBehind.md]: Hm, I didn't notice it. Emacs `markdown` mode...

File: Documentation/guides/LayoutCodeBehind.md (hunk 3 above)
Comment (grendello) [reply Documentation/guides/LayoutCodeBehind.md]: Ditto here

File: src/Mono.Android/Test/Xamarin.Android.RuntimeTests/MainActivity.cs (hunk 7 above)
Comment (grendello) [reply src/Mono.Android/Test/Xamarin.Android.RuntimeTests/MainActivity.cs]: It's a bug. `first_text_view` should be a leaf element, but not because it's a single one with this particular `id` but because it doesn't have any child elements **with** `android:id`

**Comment by grendello** [comment]
build

File: src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutCodeBehind.cs [hunk 8]
```diff
@@ -193,12 +195,40 @@ void LoadWidgets (string fileName, XmlReader reader, IXmlLineInfo lineinfo,  str
...
+
+			using (var subtree = reader.ReadSubtree ()) {
+				bool first = true;
+				while (subtree.Read ()) {
+					if (first) {
+						first = false;
+						continue;
+					}
+
+					Log.LogDebugMessage ($"Inner subtree: element == {subtree.LocalName}");
```
Comment (dellis1972) [inline src/Xamarin.Android.Build.Tasks/Tasks/GenerateLayoutCodeBehind.cs]: nice :D probably want to remote the debugging info
